### PR TITLE
Allow multiple executions of `validate()` (#672)

### PR DESCRIPTION
**Pull request by StefanoPetrilli**
#672 
I separated the different `validate()` responsibilities into other functions. It's now possible to execute `validate()` multiple times as it only performs validation.

Is this proposed solution all right? If so I will proceed with writing the tests.

### CODE REVIEW

File: include/crow/routing.h [hunk 1]
```diff
@@ -1340,7 +1363,7 @@ namespace crow
                 current_mw.merge_back(blueprint->mw_indices_);
                 for (auto& rule : blueprint->all_rules_)
                 {
-                    if (rule)
+                    if (rule && !rule->is_added())
```
Comment (cschreib) [inline include/crow/routing.h:1366]: I think you're missing a similar check there (just above):
```c++
                if (blueprint->static_dir_ == "" && blueprint->all_rules_.empty())
                {
                    std::vector<HTTPMethod> methods;
                    get_recursive_child_methods(blueprint, methods);
                    for (HTTPMethod x : methods)
                    {
                        int i = static_cast<int>(x);
                        per_methods_[i].trie.add(blueprint->prefix(), 0, blueprint->prefix().length(), i);
                    }
                }
```

It seems to me that `per_methods_[i].trie.add(...)` will keep adding routes here, if `validate_bp()` is called multiple times.

File: include/crow/app.h [hunk 2]
```diff
@@ -279,62 +279,56 @@ namespace crow
...
-                    }
-                }
-#endif
+            router_.validate_bp();
+        }
 
-                router_.validate();
-                validated_ = true;
-            }
+        //TODO(Stefano): can this be executed multiple times?
```
Comment (cschreib) [inline include/crow/app.h:306]: I don't think so :)

Comment (cschreib) [reply include/crow/app.h:306]: To clarify: I think there would have to be a flag added to make sure it isn't calling multiple times.

Comment (StefanoPetrilli) [reply include/crow/app.h:306]: Ahhh I totally forgot to check that TODO 👀

File: include/crow/routing.h (hunk 1 above)
Comment (StefanoPetrilli) [reply include/crow/routing.h:1378]: In `Blueprint` I implemented a flag like the one for `BaseRule`. It seems enough to me

Comment (StefanoPetrilli) [reply include/crow/routing.h:1378]: Take a look at my edits: in `Blueprint` I implemented a flag like the one for `BaseRule`. It seems enough to me

**Comment by cschreib** [comment]
Looks good to me. NB, I'm not a maintainer so I can't actually approve the PR for merging. But from my point of view, this would indeed fix #672.

**Comment by gittiver** [comment]
it seems a rebase onto master is needed.

**Comment by gittiver** [comment]
all checks passing, requested review by other team members.

**Review by mrozigor — APPROVED** [review]

**Pull request merged**